File: plugins/capacitor-llama-cpp/android/src/main/java/com/ethicalaiditor/plugins/llamacpp/LlamaCppPlugin.java
package com.ethicalaiditor.plugins.llamacpp;

import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.File;

/**
 * Capacitor plugin for llama.cpp inference on Android
 * 
 * This plugin wraps llama.cpp compiled for Android NDK to enable
 * local LLM inference on Android devices.
 */
@CapacitorPlugin(name = "LlamaCpp")
public class LlamaCppPlugin extends Plugin {
    private static final String TAG = "LlamaCpp";
    
    // Native library will be loaded from llama.cpp Android build
    private boolean nativeLoaded = false;
    private boolean modelLoaded = false;
    private String currentModelPath = null;
    
    // Native method declarations - implemented in llama.cpp JNI
    private native boolean nativeLoadModel(String modelPath, int contextSize, int threads, int gpuLayers, int batchSize);
    private native void nativeUnloadModel();
    private native String nativeGenerate(String prompt, int maxTokens, float temperature, float topP, int topK, float repeatPenalty, String[] stopSequences);
    private native int nativeGetTokensGenerated();
    private native long nativeGetGenerationTimeMs();
    private native long nativeGetMemoryUsage();
    
    static {
        try {
            // Load the llama.cpp native library
            System.loadLibrary("llama");
            System.loadLibrary("llamacpp_jni");
        } catch (UnsatisfiedLinkError e) {
            Log.e(TAG, "Failed to load native library: " + e.getMessage());
        }
    }
    
    @Override
    public void load() {
        super.load();
        try {
            // Verify native library is loaded
            nativeLoaded = true;
            Log.i(TAG, "LlamaCpp plugin loaded successfully");
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize LlamaCpp: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void loadModel(PluginCall call) {
        if (!nativeLoaded) {
            call.reject("Native library not loaded");
            return;
        }
        
        String modelPath = call.getString("modelPath");
        if (modelPath == null || modelPath.isEmpty()) {
            call.reject("modelPath is required");
            return;
        }
        
        // Check if file exists
        File modelFile = new File(modelPath);
        if (!modelFile.exists()) {
            call.reject("Model file not found: " + modelPath);
            return;
        }
        
        int contextSize = call.getInt("contextSize", 2048);
        int threads = call.getInt("threads", 4);
        int gpuLayers = call.getInt("gpuLayers", 0);
        int batchSize = call.getInt("batchSize", 512);
        
        // Run loading on background thread
        getActivity().runOnUiThread(() -> {
            new Thread(() -> {
                try {
                    long startTime = System.currentTimeMillis();
                    
                    boolean success = nativeLoadModel(modelPath, contextSize, threads, gpuLayers, batchSize);
                    
                    long loadTime = System.currentTimeMillis() - startTime;
                    
                    if (success) {
                        modelLoaded = true;
                        currentModelPath = modelPath;
                        
                        JSObject result = new JSObject();
                        result.put("success", true);
                        result.put("loadTimeMs", loadTime);
                        call.resolve(result);
                    } else {
                        call.reject("Failed to load model");
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error loading model: " + e.getMessage());
                    call.reject("Error loading model: " + e.getMessage());
                }
            }).start();
        });
    }
    
    @PluginMethod
    public void unloadModel(PluginCall call) {
        if (!modelLoaded) {
            JSObject result = new JSObject();
            result.put("success", true);
            call.resolve(result);
            return;
        }
        
        try {
            nativeUnloadModel();
            modelLoaded = false;
            currentModelPath = null;
            
            JSObject result = new JSObject();
            result.put("success", true);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Error unloading model: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void generate(PluginCall call) {
        if (!modelLoaded) {
            call.reject("No model loaded. Call loadModel() first.");
            return;
        }
        
        String prompt = call.getString("prompt");
        if (prompt == null || prompt.isEmpty()) {
            call.reject("prompt is required");
            return;
        }
        
        int maxTokens = call.getInt("maxTokens", 256);
        float temperature = call.getFloat("temperature", 0.7f);
        float topP = call.getFloat("topP", 0.9f);
        int topK = call.getInt("topK", 40);
        float repeatPenalty = call.getFloat("repeatPenalty", 1.1f);
        
        // Get stop sequences
        String[] stopSequences = new String[0];
        try {
            org.json.JSONArray stopArray = call.getArray("stopSequences");
            if (stopArray != null) {
                stopSequences = new String[stopArray.length()];
                for (int i = 0; i < stopArray.length(); i++) {
                    stopSequences[i] = stopArray.getString(i);
                }
            }
        } catch (Exception e) {
            // Use empty array if parsing fails
        }
        
        final String[] finalStopSequences = stopSequences;
        
        // Run generation on background thread
        new Thread(() -> {
            try {
                String generatedText = nativeGenerate(prompt, maxTokens, temperature, topP, topK, repeatPenalty, finalStopSequences);
                
                int tokensGenerated = nativeGetTokensGenerated();
                long generationTimeMs = nativeGetGenerationTimeMs();
                float tokensPerSecond = generationTimeMs > 0 ? (tokensGenerated * 1000f) / generationTimeMs : 0;
                
                JSObject result = new JSObject();
                result.put("text", generatedText);
                result.put("tokensGenerated", tokensGenerated);
                result.put("generationTimeMs", generationTimeMs);
                result.put("tokensPerSecond", tokensPerSecond);
                result.put("stoppedBySequence", false); // TODO: track this in native
                
                call.resolve(result);
            } catch (Exception e) {
                Log.e(TAG, "Error generating: " + e.getMessage());
                call.reject("Error generating: " + e.getMessage());
            }
        }).start();
    }
    
    @PluginMethod
    public void getModelInfo(PluginCall call) {
        JSObject result = new JSObject();
        result.put("isLoaded", modelLoaded);
        
        if (modelLoaded) {
            result.put("modelPath", currentModelPath);
            result.put("contextSize", 2048); // TODO: get from native
            try {
                result.put("memoryUsage", nativeGetMemoryUsage());
            } catch (Exception e) {
                result.put("memoryUsage", 0);
            }
        }
        
        call.resolve(result);
    }
    
    @PluginMethod
    public void getDeviceInfo(PluginCall call) {
        Runtime runtime = Runtime.getRuntime();
        
        JSObject result = new JSObject();
        result.put("platform", "android");
        result.put("totalMemory", runtime.maxMemory());
        result.put("availableMemory", runtime.freeMemory());
        result.put("gpuAvailable", false); // TODO: check for Vulkan support
        result.put("deviceModel", android.os.Build.MODEL);
        
        call.resolve(result);
    }
    
    @PluginMethod
    public void modelExists(PluginCall call) {
        String path = call.getString("path");
        if (path == null) {
            call.reject("path is required");
            return;
        }
        
        File file = new File(path);
        JSObject result = new JSObject();
        result.put("exists", file.exists());
        call.resolve(result);
    }
    
    @PluginMethod
    public void getRecommendedConfig(PluginCall call) {
        Runtime runtime = Runtime.getRuntime();
        long maxMemory = runtime.maxMemory();
        int availableProcessors = runtime.availableProcessors();
        
        // Estimate based on available memory
        int maxContextSize = maxMemory > 4L * 1024 * 1024 * 1024 ? 4096 : 2048;
        int recommendedThreads = Math.min(availableProcessors, 8);
        long maxModelSize = maxMemory / 2; // Use at most half of available memory
        
        JSObject result = new JSObject();
        result.put("maxContextSize", maxContextSize);
        result.put("recommendedThreads", recommendedThreads);
        result.put("gpuLayersSupported", 0); // Conservative default
        result.put("maxModelSizeBytes", maxModelSize);
        
        call.resolve(result);
    }
}

File: plugins/capacitor-llama-cpp/ios/Plugin/LlamaCppPlugin.swift
import Foundation
import Capacitor

/**
 * Capacitor plugin for llama.cpp inference on iOS
 *
 * This plugin wraps llama.cpp compiled for iOS to enable
 * local LLM inference on iOS devices using Metal for GPU acceleration.
 */
@objc(LlamaCppPlugin)
public class LlamaCppPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LlamaCppPlugin"
    public let jsName = "LlamaCpp"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "loadModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unloadModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getModelInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDeviceInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "modelExists", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getRecommendedConfig", returnType: CAPPluginReturnPromise)
    ]
    
    private var llamaContext: LlamaContext?
    private var isModelLoaded = false
    private var currentModelPath: String?
    
    // MARK: - Plugin Methods
    
    @objc func loadModel(_ call: CAPPluginCall) {
        guard let modelPath = call.getString("modelPath") else {
            call.reject("modelPath is required")
            return
        }
        
        // Check if file exists
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: modelPath) else {
            call.reject("Model file not found: \(modelPath)")
            return
        }
        
        let contextSize = call.getInt("contextSize") ?? 2048
        let threads = call.getInt("threads") ?? 4
        let gpuLayers = call.getInt("gpuLayers") ?? 0
        let batchSize = call.getInt("batchSize") ?? 512
        
        // Load on background queue
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            let startTime = CFAbsoluteTimeGetCurrent()
            
            do {
                // Initialize llama context with model
                self.llamaContext = try LlamaContext(
                    modelPath: modelPath,
                    contextSize: contextSize,
                    threads: threads,
                    gpuLayers: gpuLayers,
                    batchSize: batchSize
                )
                
                let loadTime = (CFAbsoluteTimeGetCurrent() - startTime) * 1000
                
                self.isModelLoaded = true
                self.currentModelPath = modelPath
                
                DispatchQueue.main.async {
                    call.resolve([
                        "success": true,
                        "loadTimeMs": Int(loadTime)
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    call.reject("Failed to load model: \(error.localizedDescription)")
                }
            }
        }
    }
    
    @objc func unloadModel(_ call: CAPPluginCall) {
        llamaContext = nil
        isModelLoaded = false
        currentModelPath = nil
        
        call.resolve(["success": true])
    }
    
    @objc func generate(_ call: CAPPluginCall) {
        guard isModelLoaded, let context = llamaContext else {
            call.reject("No model loaded. Call loadModel() first.")
            return
        }
        
        guard let prompt = call.getString("prompt") else {
            call.reject("prompt is required")
            return
        }
        
        let maxTokens = call.getInt("maxTokens") ?? 256
        let temperature = call.getFloat("temperature") ?? 0.7
        let topP = call.getFloat("topP") ?? 0.9
        let topK = call.getInt("topK") ?? 40
        let repeatPenalty = call.getFloat("repeatPenalty") ?? 1.1
        let stopSequences = call.getArray("stopSequences", String.self) ?? []
        
        // Generate on background queue
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            let startTime = CFAbsoluteTimeGetCurrent()
            
            do {
                let result = try context.generate(
                    prompt: prompt,
                    maxTokens: maxTokens,
                    temperature: temperature,
                    topP: topP,
                    topK: topK,
                    repeatPenalty: repeatPenalty,
                    stopSequences: stopSequences
                )
                
                let generationTime = (CFAbsoluteTimeGetCurrent() - startTime) * 1000
                let tokensPerSecond = generationTime > 0 ? Double(result.tokensGenerated) * 1000.0 / generationTime : 0
                
                DispatchQueue.main.async {
                    call.resolve([
                        "text": result.text,
                        "tokensGenerated": result.tokensGenerated,
                        "generationTimeMs": Int(generationTime),
                        "tokensPerSecond": tokensPerSecond,
                        "stoppedBySequence": result.stoppedBySequence
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    call.reject("Generation failed: \(error.localizedDescription)")
                }
            }
        }
    }
    
    @objc func getModelInfo(_ call: CAPPluginCall) {
        var result: [String: Any] = ["isLoaded": isModelLoaded]
        
        if isModelLoaded {
            result["modelPath"] = currentModelPath
            result["contextSize"] = llamaContext?.contextSize ?? 2048
            result["memoryUsage"] = llamaContext?.memoryUsage ?? 0
        }
        
        call.resolve(result)
    }
    
    @objc func getDeviceInfo(_ call: CAPPluginCall) {
        let processInfo = ProcessInfo.processInfo
        let device = UIDevice.current
        
        call.resolve([
            "platform": "ios",
            "totalMemory": processInfo.physicalMemory,
            "availableMemory": getAvailableMemory(),
            "gpuAvailable": true, // Metal is available on all modern iOS devices
            "deviceModel": device.model
        ])
    }
    
    @objc func modelExists(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("path is required")
            return
        }
        
        let exists = FileManager.default.fileExists(atPath: path)
        call.resolve(["exists": exists])
    }
    
    @objc func getRecommendedConfig(_ call: CAPPluginCall) {
        let totalMemory = ProcessInfo.processInfo.physicalMemory
        let availableMemory = getAvailableMemory()
        
        // Estimate based on device memory
        let maxContextSize = totalMemory > 4 * 1024 * 1024 * 1024 ? 4096 : 2048
        let recommendedThreads = min(ProcessInfo.processInfo.processorCount, 8)
        let maxModelSize = availableMemory / 2 // Use at most half of available memory
        
        // Metal GPU layers - conservative estimate based on device
        let gpuLayersSupported = totalMemory > 6 * 1024 * 1024 * 1024 ? 32 : 16
        
        call.resolve([
            "maxContextSize": maxContextSize,
            "recommendedThreads": recommendedThreads,
            "gpuLayersSupported": gpuLayersSupported,
            "maxModelSizeBytes": maxModelSize
        ])
    }
    
    // MARK: - Helper Methods
    
    private func getAvailableMemory() -> UInt64 {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size) / 4
        
        let result = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: 1) {
                task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), $0, &count)
            }
        }
        
        if result == KERN_SUCCESS {
            let usedMemory = info.resident_size
            let totalMemory = ProcessInfo.processInfo.physicalMemory
            return totalMemory - usedMemory
        }
        
        return ProcessInfo.processInfo.physicalMemory / 2 // Fallback estimate
    }
}
